src/components/lab/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,320 @@
+    );
+
+    React.useEffect(() => {
+        const stringPropsValue = getStringValue(value);
+        setInputValue((currentInputValue) => {
+            if (!areStringRepresentationOfNumbersEqual(currentInputValue, stringPropsValue)) {
+                return stringPropsValue;
+            }
+            return currentInputValue;
+        });
+    }, [value]);
+
+    const clamp = true;
+    const [step, setStep] = React.useState(baseStep);
+    const [active, setActive] = React.useState(false);
+
+    const state = getInputControlState(validationState);
+
+    const canIncrementNumber =
+        clampToNearestStepValue({value: value + step, step, min, max}) > value;
+
+    const canDecrementNumber =
+        clampToNearestStepValue({value: value - step, step, min, max}) < value;
+
+    const innerControlRef = React.useRef<HTMLInputElement>(null);
+    const fieldRef = useFormResetHandler({
+        initialValue: defaultValue,
+        onReset: onUpdate ?? voidFunction,
+    });
+    const handleRef = useForkRef(props.controlRef, innerControlRef, fieldRef);
+
+    const handleIncrement = () => {
+        if (canIncrementNumber) {
+            onUpdate?.(clampToNearestStepValue({value: value + step, step: baseStep, min, max}));
+        }
+    };
+
+    const handleDecrement = () => {
+        if (canDecrementNumber) {
+            onUpdate?.(clampToNearestStepValue({value: value - step, step: baseStep, min, max}));
+        }
+    };
+
+    const handleWheel: React.WheelEventHandler<HTMLInputElement> = (e) => {
+        const delta = e.shiftKey ? e.deltaX : e.deltaY;
+        if (delta > 0) {
+            handleIncrement();
+        } else if (delta < 0) {
+            handleDecrement();
+        }
+    };
+
+    const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+        if (e.key === KeyCode.SHIFT) {
+            setStep(baseStep * shiftMultiplier);
+        } else if (e.key === KeyCode.ARROW_DOWN) {
+            handleDecrement();
+        } else if (e.key === KeyCode.ARROW_UP) {
+            handleIncrement();
+        }
+        onKeyDown?.(e);
+    };
+
+    const handleKeyUp: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+        if (e.key === KeyCode.SHIFT) {
+            setStep(baseStep);
+        }
+        onKeyUp?.(e);
+    };
+
+    const handleFocus: React.FocusEventHandler<HTMLInputElement> = (e) => {
+        setActive(true);
+        onFocus?.(e);
+    };
+
+    const handleBlur: React.FocusEventHandler<HTMLInputElement> = (e) => {
+        setActive(false);
+        if (clamp) {
+            const clampedValue = clampToNearestStepValue({
+                value,
+                step: baseStep,
+                min,
+                max,
+            });
+            if (value !== clampedValue) {
+                onUpdate?.(clampedValue);
+            }
+        }
+        onBlur?.(e);
+    };
+
+    const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
```

Comment:
We can only use `onUpdate` for updating the value. `onChange` handler is redundant